bundles/specmate-migration-test/src/com/specmate/migration/test/AddAttributeTest.java

```diff
@@ -0,0 +1,229 @@
+		properties.put(CDOPersistenceConfig.KEY_JDBC_CONNECTION, "jdbc:h2:mem:testdb;DB_CLOSE_DELAY=-1");
+		//properties.put(CDOPersistenceConfig.KEY_JDBC_CONNECTION, "jdbc:h2:./database/specmate");
+		properties.put(CDOPersistenceConfig.KEY_REPOSITORY_NAME, "testrepo");
+		properties.put(CDOPersistenceConfig.KEY_RESOURCE_NAME, "r1");
+		properties.put(CDOPersistenceConfig.KEY_USER_RESOURCE_NAME, "r2");	
+		configurationAdmin.getConfiguration(CDOPersistenceConfig.PID).update(properties);
+		
+		baselineModelController = new ServiceController<>(context);
+		baselineModelController.register(IPackageProvider.class, BaselineModelProviderImpl.class, null);
+		attributeAddedModelController = new ServiceController<>(context);
+		attributeAddedModelController.register(IPackageProvider.class, AttributeAddedModelProviderImpl.class, null);
+		persistencyServiceController = new ServiceController<>(context); 
+		migratorService = getMigratorService();
+				
+		activatePersistency(baselineModelController.getService());
+		addBaselinedata();
+		deactivatePersistency();
+	}
+	
+	@AfterClass
+	public static void shutdown() {
+		baselineModelController.unregister();
+		attributeAddedModelController.unregister();
+	}
+	
+	private static void activatePersistency(IPackageProvider p) throws Exception {
+		persistencyServiceController.register(IPersistencyService.class, CDOPersistencyService.class, null);
+		persistencyService = persistencyServiceController.getService();
+		CDOPersistencyService cdop = (CDOPersistencyService) persistencyService;
+		startSupportServices(cdop, p);
+		cdop.activate();
+	}
+	
+	private static void deactivatePersistency() {
+		CDOPersistencyService cdop = (CDOPersistencyService) persistencyService;
+		cdop.deactivate();
+		persistencyServiceController.unregister();
+	}
+	
+	@Test 
+	public void testNeedsMigration() throws Exception {
+		activatePersistency(baselineModelController.getService());
+		assertFalse(migratorService.needsMigration());
+		migratorService.setModelProviderService(attributeAddedModelController.getService());
+		assertTrue(migratorService.needsMigration());
+		deactivatePersistency();
+	}
+	
+	@Test
+	public void doMigration() throws Exception {
+		activatePersistency(baselineModelController.getService());
+		ITransaction transaction = persistencyService.openTransaction();
+		Resource resource = transaction.getResource();
+		EObject root = SpecmateEcoreUtil.getEObjectWithName("root", resource.getContents());
+		assertNotNull(root);
+		EObject diagram = SpecmateEcoreUtil.getEObjectWithName("d0", root.eContents());
+		assertNotNull(diagram);
+		assertNull(SpecmateEcoreUtil.getAttributeValue(root, "id", String.class));
+		assertNull(SpecmateEcoreUtil.getAttributeValue(diagram, "id", String.class));
+		transaction.close();
+		deactivatePersistency();
+		
+		migratorService.setModelProviderService(attributeAddedModelController.getService());
+		migratorService.doMigration();
```

Comment:
why do you need to tirgger this manually? Should'nt the persistency service detect the need for migration and trigger the migration itself?

Reply:
This is due to the the following:
1) I need to control the life-cycle of the CDOPersistencyService in order to be able to (a) generate the database and data, and (b) run the migration in ONE execution of the test cases. I must be able to replace the data model at runtime and this is only possible if I can shut down the restart the persistency service.
2) When I start the persistency service manually (see ServiceController), the CDOPersistencyService.activate() method is NOT called automatically by the OSGi framework. The activate method of the persistency service performs the migration, which I cannot call since it expects a "Config" object which is private to the CDOPersistencyService class. Hence, as a workaround, I have added an activate method that I can call from the test. This is ugly, so if there is another way to solve this, without changing business code, I'm open to suggestions.

Reply:
ok, so then let's leave like this for now (maybe you could add a code comment)

Reply:
There is already a comment in CDOPersistencyService that explains the purpose of the added activate method. I've added there what we need to consider when fixing this workaround.

Reply:
sorry for being picky :-) I ment a comment in the test procedure (line 120)

Reply:
;) done